primitives/src/parachain.rs

```diff
@@ -792,6 +779,101 @@ pub mod id {
 	pub const PARACHAIN_HOST: ApiId = *b"parahost";
 }
 
+/// This helper trait ensures that we can encode Statement as CompactStatement,
+/// and anything as itself.
+///
+/// This resembles `parity_scale_codec::EncodeLike`, but it's distinct:
+/// EncodeLike is a marker trait which asserts at the typesystem level that
+/// one type's encoding is a valid encoding for another type. It doesn't
+/// perform any type conversion when encoding.
+///
+/// This trait, on the other hand, provides a method which can be used to
+/// simultaneously convert and encode one type as another.
+pub trait EncodeAs<T> {
+	/// Convert Self into T, then encode T.
+	///
+	/// This is useful when T is a subset of Self, reducing encoding costs;
+	/// its signature also means that we do not need to clone Self in order
+	/// to retain ownership, as we would if we were to do
+	/// `self.clone().into().encode()`.
+	fn encode_as(&self) -> Vec<u8>;
+}
+
+impl<T: Encode> EncodeAs<T> for T {
+	fn encode_as(&self) -> Vec<u8> {
+		self.encode()
+	}
+}
+
+/// A signed type which encapsulates the common desire to sign some data and validate a signature.
+///
+/// Note that the internal fields are not public; they are all accessable by immutable getters.
+/// This reduces the chance that they are accidentally mutated, invalidating the signature.
+#[derive(Debug, Clone, PartialEq, Encode, Decode)]
+pub struct Signed<Payload, RealPayload = Payload> {
+	/// The payload is part of the signed data. The rest is the signing context,
+	/// which is known both at signing and at validation.
+	payload: Payload,
+	/// The index of the validator signing this statement.
+	validator_index: ValidatorIndex,
+	/// The signature by the validator of the signed payload.
+	signature: ValidatorSignature,
+	/// This ensures the real payload is tracked at the typesystem level.
+	real_payload: sp_std::marker::PhantomData<RealPayload>,
+}
+
+// We can't bound this on `Payload: Into<RealPayload>` beacuse that conversion consumes
+// the payload, and we don't want that. We can't bound it on `Payload: AsRef<RealPayload>`
+// because there's no blanket impl of `AsRef<T> for T`. In the end, we just invent our
+// own trait which does what we need: EncodeAs.
+impl<Payload: EncodeAs<RealPayload>, RealPayload: Encode> Signed<Payload, RealPayload> {
+	fn payload_data(payload: &Payload, context: SigningContext) -> Vec<u8> {
+		(payload.encode_as(), context).encode()
```

Comment:
Note that this is different than the encoding before and in fact is two encodings.

`(Vec<u8>, X)` is different than `(Y, X)`.

Reply:
Getting around this requires some serious trait-juggling to avoid clones.

Reply:
I thought that in either case, scale just appended them? Of course you're right that those are different types, though.

I'll think on this some more, but no straightforward solution is coming to mind yet. The intent was to separate the context-appending from the encoding stage, but maybe that isn't possible while retaining compatibility.

Reply:
Possibly this is where `EncodeLike` comes into play, to decode the `(Vec<u8>, X)` into `(Y, X)`.

Reply:
I posted a playground to fix this issue. For a tuple (A, B), SCALE will append `A.encode()` and `B.encode()`.

However you are doing `A.encode().encode()`

Reply:
Thanks for the clear explanation! I'd missed that issue. I think it's now fixed, so I'm going to resolve this.